### ISSUE TITLE
Return a Promise instead of accepting callbacks

### ISSUE THREAD

**Issue by AurelioDeRosa** [michael/github#233]
Hi.

Considering the current approach used by many libraries and JavaScript APIs, I think it'd be beneficial for the library's methods to return a Promise instead of accepting a callback. Using Promises is more flexible and this approach gives developers more control about what happens when a request is performed.

**Comment by aendrew** [comment]
Not sure I agree; the callback style is still used pretty extensively throughout Nodeland and changing them to return promises is a pretty major API change. Plus [Octokat](https://github.com/philschatz/octokat.js) already does that, and you can use promises with Github.js by using [Q.denodeify](https://github.com/kriskowal/q#adapting-node).

If both the callback style and promises are possible, then go for it. If it means breaking backwards compatibility, it's probably worth fixing everything else first and doing this for the next major version. Am pretty open either way, though.

**Comment by AurelioDeRosa** [comment]
I guess that from the point of view of the library it doesn't make much difference but it does for the users. The library always performs single, simple calls. So, it isn't a big deal for the library. The benefit is usually for users of the library when they have to run multiple actions one after the other or in parallel. In this case the code with callbacks becomes a mess.

In regard of your question, I think it needs a big change. Therefore, I'm fine in postponing it. However, I feel this is really something we should do.

**Comment by michael** [comment]
Yeah, that's where javascript hurts still. ;) Promises are definitely making things better. However, still there's some overhead for the caller and a methods that returns promises are not so easy to write.

ES7 async/await syntax will make this finally nice and clear to write. It's already possible with ES6 and generators (but not so easy to understand).

Read this: https://jakearchibald.com/2014/es7-async-functions/

I wonder if Github.js would be suitable to be written in ES6 and compiled using Babel. I'm usually against this as each source compilation step makes things harder to debug and slower to develop. However given this is a like a single-file repo, this might not hurt much, debugging is still possible thanks to source maps.

Just mentioning this, as you may consider upgrading the lib to ES6 anyway at some point. And this would be a more sustainable solution than Promises probably. Babel has experimental support for ES7... in case you want to play with it a bit.

**Comment by aendrew** [comment]
Hmm, for realz it would be rad to rewrite the lib as an ES2015 module. We could definitely do it as an ES2015 class for version 1.x, leaving the 0.1.x branch as the existent ES5 library. It would also really move the library forward into being more future-facing instead of just playing catch-up. I'm on board. 

TBTH I'd say for legacy reasons we should get through 0.10.8 at least before considering that, making sure that the existing releases are more solid than they've been. But it's definitely an idea as we move into the future, and I'm completely down for doing so. 

re: promises in ES6 — you can [already do them with Babel](https://babeljs.io/docs/learn-es2015/#promises), though it requires the polyfill in the browser. If that's too much, I really like [kriskowal/q](https://github.com/kriskowal/q), and wouldn't feel too bad about adding it as a dependency in Node (even if added to a runtime via Browserify, it's only 3.5kb minified). 

FWIW I didn't mean to pour cold water on the idea of promises so much as prioritise the outstanding queue first — I really dig the patterns you can do with promises, and if adding them won't increase the weight of the library too much, it should be able to do that. What if we used a pattern whereby it returns a promise if there's no callback supplied? Not entirely sure what is the most intuitive. Definitely something for discussion, though.

**Comment by AurelioDeRosa** [comment]
This can be closed as we'll manage async functions with Babel. At this stage, I'd probably say that we'll skip the Promises land altogether and move directly into async/await.

**Comment by jacob-israel-turner** [comment]
Is this being tracked in a separate issue?  I'm currently using another (abandoned) Github project, and I'm looking for a library that uses promises.  I have experience using Babel, Promises, and a little bit with async/await.  If I can help in any way, I'm more than willing to.

**Comment by AurelioDeRosa** [comment]
Hi.

We've currently put Promises aside. We're also focusing of fixing bugs and extending the features provided. So, if you wish to help try to fix one of the issues reported.

**Comment by jacob-israel-turner** [comment]
@AurelioDeRosa would you be opposed to me submitting a backwards-compatible PR introducing promises?  This is the biggest feature I'd be interested in.

**Comment by AurelioDeRosa** [comment]
I won't be opposed as it's your time and you can do whatever you wish. However, considering our plans to change the library, your request will need a lot of future rebase and work. Anyway, I'd like to hear the opinion of @aendrew, @michael, and @clayreimann on this matter.

**Comment by jacob-israel-turner** [comment]
:+1: I'll start investigating the codebase and I'll await feedback before submitting any PRs.
